### PR TITLE
CL/HIER: check number of TLs per SBGP

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Add check to CL/HIER to prevent SBGP TL list overflow.

## Why ?
For some systems CL/HIER might try to use more than 4 TLs per SBGP. 4 is compile time constant. Fixes bug https://redmine.mellanox.com/issues/3767158

## How ?
Ignore TLs if the list already full. Filter out TLs if we know in advance that SBGP size is not supported.

### CODE REVIEW

File: test/mpi/main.cc [hunk 1]
```diff
@@ -574,15 +574,17 @@ void ProcessArgs(int argc, char** argv)
 
 int main(int argc, char *argv[])
 {
-    int failed                                                          = 0;
-    int total_done_skipped_failed[ucc_ilog2(UCC_COLL_TYPE_LAST) + 1][4] = {0};
+    int failed = 0;
+    int total_done_skipped_failed[ucc_ilog2(UCC_COLL_TYPE_LAST) + 1][4];
```
Comment (janjust) [inline test/mpi/main.cc:578]: I guess unnecessary change for the bug, but does remove a compiler warning.

**Review by janjust — APPROVED** [review]

File: src/components/cl/hier/cl_hier_team.c [hunk 2]
```diff
@@ -74,18 +79,51 @@ UCC_CLASS_INIT_FUNC(ucc_cl_hier_team_t, ucc_base_context_t *cl_context,
             hs->n_tls = 0;
             tls       = &lib->cfg.sbgp_tls[i].array;
```
Comment (bureddy) [inline src/components/cl/hier/cl_hier_team.c:80]: qq: is orders of TLS in this array always will be same? do you see any issues if there is a mismatched list of TLS selected for team on different ranks?

**Review by bureddy — APPROVED** [review]

File: test/mpi/main.cc (hunk 1 above)
Comment (Sergei-Lebedev) [reply test/mpi/main.cc:578]: you are right, it just fixes gcc 4.8.5 warning when build ucc_test_mpi

File: src/components/cl/hier/cl_hier_team.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/components/cl/hier/cl_hier_team.c:80]: it matters, there is separate fix for this https://github.com/openucx/ucc/pull/717, but it was rejected since the bigger change in UCC is needed to detect mismatches for every CL and core

File: src/components/cl/hier/cl_hier_team.c [hunk 3]
```diff
@@ -74,18 +79,51 @@ UCC_CLASS_INIT_FUNC(ucc_cl_hier_team_t, ucc_base_context_t *cl_context,
...
             for (j = 0; j < tls->count; j++) {
+                if (hs->n_tls == CL_HIER_MAX_SBGP_TLS) {
+                    cl_debug(cl_context->lib,
+                             "skipping tl context %s for %s sbgp: "
+                             "max number of TLs per SBGP is reached",
+                             tls->names[j], ucc_sbgp_str(hs->sbgp_type));
+                    continue;
```
Comment (samnordmann) [inline src/components/cl/hier/cl_hier_team.c:87]: break ?

**Review by samnordmann — APPROVED** [review]

File: src/components/cl/hier/cl_hier_team.c (hunk 3 above)
Comment (Sergei-Lebedev) [reply src/components/cl/hier/cl_hier_team.c:87]: we reached max and could break, but we want to print in debug msg what TLs were skipped

**Comment by artemry-nv** [comment]
bot:retest
